xdsl/utils/lexer.py

```diff
@@ -122,3 +134,386 @@ def print_with_context(self, msg: str | None = None) -> str:
+        """
+        Parsed the following prefixed identifiers:
+        ```
+        hash-ident  ::= `#` suffix-id
+        percent-ident  ::= `%` suffix-id
+        caret-ident  ::= `^` suffix-id
+        exclamation-ident  ::= `!` suffix-id
+        ```
+        with 
+        ```
+        suffix-id = (digit+ | (letter|[$._-]) (letter|[$._-]|digit)*)'''
+        ```
+
+        The first character is expected to have already been parsed.
+        """
+        assert self.pos != 0, "First prefixed identifier character must have been parsed"
+        first_char = self.input.at(self.pos - 1)
+        if first_char == '#':
+            kind = Token.Kind.HASH_IDENT
+        elif first_char == '!':
+            kind = Token.Kind.EXCLAMATION_IDENT
+        elif first_char == '^':
+            kind = Token.Kind.CARET_IDENT
+        else:
+            assert first_char == '%', "First prefixed identifier character must have been parsed correctly"
+            kind = Token.Kind.PERCENT_IDENT
+
+        punctuation = ['$', '.', '_', '-']
+
+        current_char = self._get_chars()
+        if current_char is None:
+            raise ParseError(
+                Span(start_pos, start_pos + 1, self.input),
+                "'first_char' is expected to follow with an identifier.")
+
+        if current_char.isdigit():
+            # digit+ case
+            self._consume_while(lambda c: c.isdigit())
+        elif current_char.isalpha() or current_char in punctuation:
+            # (letter|[$._-]) (letter|[$._-]|digit)* case
+            self._consume_while(lambda c: c.isalnum() or c in punctuation)
+        else:
+            raise ParseError(
+                Span(start_pos, start_pos + 1, self.input),
+                "Character expected to follow with an identifier.")
+
+        return self._form_token(kind, start_pos)
+
+    def _lex_string_literal(self, start_pos: int) -> Token:
+        """
+        Lex a string literal.
+        The first character `"` is expected to have already been parsed.
+        """
+
+        while self._is_in_bounds():
+            current_char = self._get_chars()
+
+            # end of string literal
+            if current_char == '"':
+                return self._form_token(Token.Kind.STRING_LIT, start_pos)
+
+            # newline character in string literal (not allowed)
+            if current_char in ['\n', '\v', '\f']:
+                raise ParseError(
+                    Span(start_pos, self.pos, self.input),
+                    "Newline character not allowed in string literal.")
+
+            # escape character
+            # TODO: handle unicode escape
+            if current_char == '\\':
+                escaped_char = self._get_chars()
+                if escaped_char not in ['"', '\\', 'n', 't']:
+                    raise ParseError(Span(self.pos - 1, self.pos, self.input),
+                                     "Unknown escape in string literal.")
```

Comment:
I'd love to remove the dependency on `ast`

Reply:
I use `StringLiteral` for now, I'll see later how I actually handle values in `Span` (for instance, int literal may carry an int instead of a string?)